amaranth/ml/train.py

```diff
@@ -52,24 +57,69 @@ def main():
       low_calorie_threshold=amaranth.LOW_CALORIE_THRESHOLD,
       high_calorie_threshold=amaranth.HIGH_CALORIE_THRESHOLD)
 
+  # Normalize input strings
+  # Step 1: convert strings to lowercase
+  # Step 2: filter out characters present in DISH_NAME_FILTERS
+  # Step 3: re-combine characters back to a normal string
+  calorie_data['description'] = calorie_data['description'].apply(
+      lambda desc: desc.lower(),  # Step 1
+  ).apply(
+      lambda desc: [char for char in desc
+                    if char not in DISH_NAME_FILTERS],  # Step 2
+  ).apply(
+      ''.join,  #Step 3
+  )
+
   # Do some preprocessing and calculations for encoding
-  corpus = calorie_data['description'].str.replace(',', '').str.lower()
+  corpus = calorie_data['description']
   vocab_size = lib.num_unique_words(corpus)
   tokenized_corpus = corpus.map(lambda desc: desc.split(' '))
   max_corpus_length = lib.max_sequence_length(tokenized_corpus)
 
-  # Create TextVectorization layer
-  vectorize_layer = keras.layers.experimental.preprocessing.TextVectorization(
-      max_tokens=vocab_size,
-      output_sequence_length=max_corpus_length,
-  )
-
-  vectorize_layer.adapt(calorie_data['description'].to_numpy())
+  # Encode 'description' into new column 'input'
+  calorie_data['tokenized'] = calorie_data.apply(
+      lambda row: text.one_hot(row['description'], vocab_size), axis=1)
+
+  # Create tokenizer to encode input text
+  tokenizer = dict()  # Dictionary mapping each unique word to a unique int
+  tokenizer_cnt = dict()  # Counts appearances of each word in tokenizer
+  for _, row in calorie_data.iterrows():
+    for idx, word in enumerate(row['description'].split()):
+      tokenizer[word] = row['tokenized'][idx]
+
+      if word in tokenizer_cnt:
+        tokenizer_cnt[word] += 1
+      else:
+        tokenizer_cnt[word] = 1
+
+  # Only 'remember' words that appear at least 3 times
+  for word, cnt in tokenizer_cnt.items():
+    if cnt < MIN_TOKEN_APPEARANCE:
+      del tokenizer[word]
+
+  # The empty string denotes words that are OOV (out-of-vocabulary)
+  # It is equal to vocab_size because all values in the tokenizer should be less
+  # than vocab_size
+  tokenizer[''] = 0
```

Comment:
1) if an item is actually OOV, might be worth calling it OOV i.e.
``` tokenizer['OOV'] = 0 ```
2) Just curious, but why are you assigning 0 as the value?

Reply:
1. Good point, will change that!
2. As my (accidentally outdated) comment states, it used to be equal to `vocab_size` but that caused issues because the model expects values strictly less than `vocab_size`. I could tell the model to expect values thru `vocab_size+1`, but Tensorflow.js currently has no way to determine what `vocab_size` is/was

Reply:
Right. I think when you create the model below, you also need to make sure the embedding sizes are correct since technically you are adding one additional "word". Also, in general, it might be useful to just define your OOV token as 0; that way, the vocab_size variable doesn't apply on the JS side of the equation. 
```
keras.layers.Embedding(vocab_size+1, int((vocab_size+1)**(0.25) ...)
```

Reply:
Point 1 fixed in https://github.com/googleinterns/amaranth/pull/35/commits/4e02b52d0afefda12671b1113402764622c952f2.

> Right. I think when you create the model below, you also need to make sure the embedding sizes are correct since technically you are adding one additional "word".
> 
> ```
> keras.layers.Embedding(vocab_size+1, int((vocab_size+1)**(0.25) ...)
> ```

Fixed in https://github.com/googleinterns/amaranth/pull/35/commits/9eb700437b03de7bac9a812bd470f2c5fdd99838